### DISCUSSION TITLE
Static linking in Appimage

### DISCUSSION THREAD

**Discussion by robertosw** [tauri-apps/tauri#8367] [Ideas]
Would it be possible to statically link all used libraries when building an AppImage?
After all, the main use of the AppImage is to enable it to run on all linux distros. 
Currently the AppImage only works in Debian-like distros, because all library paths are dynamically linked but with fixed paths.
This results in problems wenn running the AppImage on NixOS where all libraries are in the nix store, so the paths from the AppImage are never found.

I tried setting my rust target to linux-musl, but some transitive dependencies dont support this target.

**Comment by FabianLars** [comment]
This is super hard and at this point i doubt it's really possible with our current stack. Maybe once we have a servo integration that could be statically linked 🤔 

Also, can you share the offending paths or some errors? The appimages are supposed to work on non-debian-like distros too, we even patch the webkitgtk .so binary for that 😅

**Reply by robertosw** [reply]
I built my AppImage inside a docker container running debain. Everything works if I start this inside the container.
Starting the file on my host, results in this:
```Console
$ appimage-run myapp_0.1.7_amd64.AppImage 
myapp_0.1.7_amd64.AppImage installed in /home/robert/.cache/appimage-run/8e9b9bb917f764c2e1dfab1a24ac8e51c5f3f640ecf7bffc4780614a613698d3
myapp: error while loading shared libraries: libthai.so.0: cannot open shared object file: No such file or directory
```

libthai has to be some transitive dependency, because my project doesnt use that (afaik)

**Reply by FabianLars** [reply]
Ah right, the classic https://github.com/tauri-apps/tauri/issues/4930 - i was hoping this would magically resolve itself upstream but maybe not :/

**Reply by robertosw** [reply]
Oh okay good to that this is a known bug :D

> Thanks! I still have a couple of follow up questions. If this error happens on your system:
>     Which distro are you using?
>    what is your display language?
>    what is your input language?
>    I'm asking because this doesn't happen on english & german systems.

This is from the issue you linked. Im also on nixos, with german as the main language and keyboard layout

**Reply by FabianLars** [reply]
You can basically ignore that part. When i asked this i didn't know we were talking about nixos and i only tested it on ubuntu&fedora where this isn't an issue.

**Reply by robertosw** [reply]
Oh okay, but is there any solution to this problem? 
I already have the libthai package installed but this error doesnt go away. 
To be fair, I am not entirely sure that the package also installed the library for every other app to use

**Reply by FabianLars** [reply]
> Oh okay, but is there any solution to this problem?

Not reallyyy. There's the upstream discussion about a solution (<https://github.com/tauri-apps/tauri/issues/4930>) but nobody's really confident 😅

**Reply by robertosw** [reply]
Is there any use installing the missing lib on my device?  Or is it impossible to tell the AppImage where to search for that lib?

**Reply by FabianLars** [reply]
spawning the appimage like so may work `LD_LIBRARY_PATH="/path/to/libdir:$LD_LIBRARY_PATH" myapp.AppImage` may work, assuming the system you run on has the required libraries installed.

**Comment by ilyagr** [comment]
This seems obvious in retrospect, but static linking would increase the size of the static binary *dramatically*. I haven't gotten it to work, but I know that installing the `libwebkit2gtk-4.0-37` runtime dependency and all of its dependencies takes up **400MB** on Ubuntu. I found this out in https://github.com/lima-vm/lima/discussions/2215#discussioncomment-8509432 (where the dynamic binary in question is a small Tauri program, and I truly installed it all from scratch since I was trying to run it all on a different architecture).
